### PR TITLE
Deterministic errors

### PR DESCRIPTION
**Pull request by leoyvens**
This marks as determinstic few common reasons for subgraph failures, for example:
- An `unreachable` wasm trap which is caused by an array access out of bounds.
- An `abort` which AssemblyScript uses for multiple reasons including failed assert.
- Some uncontroversial host export erros such as division by zero.

There may be more cases we'd like to cover, but those can be added as long as we consider it a breaking change and bump the `specVersion` (or `apiVersion`?). Deterministic errors are flagged by a boolean in the metadata which can be accessed through the indexing status API.

### CODE REVIEW

File: core/src/subgraph/instance_manager.rs [hunk 1]
```diff
@@ -595,6 +596,30 @@ where
     }
 }
 
+#[derive(thiserror::Error, Debug)]
+enum ProcessBlockError {
```
Comment (lutter) [inline core/src/subgraph/instance_manager.rs:600]: To me, `BlockProcessingError` or even `IndexingError` would be clearer than `ProcessBlockError`

File: core/src/subgraph/instance_manager.rs [hunk 2]
```diff
@@ -663,7 +688,12 @@ where
     .await
     {
         Ok(block_state) => block_state,
-        Err(MappingError::Unknown(e)) => return Err(e.compat_err().into()),
+        Err(MappingError::Unknown(e)) => {
+            return Err(CancelableError::Error(ProcessBlockError::Unknown(e)))
```
Comment (lutter) [inline core/src/subgraph/instance_manager.rs:692]: Is there any advantage to have separate `MappingError` and `ProcessBlockError`? Maybe we should just elevate `MappingError` to be used throughout?

**Review by lutter — APPROVED** [review]

File: core/src/subgraph/instance_manager.rs (hunk 1 above)
Comment (leoyvens) [reply core/src/subgraph/instance_manager.rs]: Renamed to `BlockProcessingError`.

File: core/src/subgraph/instance_manager.rs (hunk 2 above)
Comment (leoyvens) [reply core/src/subgraph/instance_manager.rs]: `ProcessBlockError` doesn't have the `PossibleReorg` variant.

Comment (lutter) [reply core/src/subgraph/instance_manager.rs]: But for that one, we know that it's not deterministic for sure.

Comment (leoyvens) [reply core/src/subgraph/instance_manager.rs]: I just see it as an advantage to have a fine grained types with less variants, and I don't see what code would be greatly simplified by having a single type.

**Pull request merged**